File: Data Structures & Algorithms/count-number-of-islands/submission-3.java
class Solution {

    private static final int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    public int numIslands(char[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        int islands = 0;

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == '1') {
                    dfs(row, col, grid);
                    islands++;
                }
            }
        }

        return islands;
    }

    private void dfs(int row, int col, char[][] grid) {
        if(row < 0 || col < 0 || row >= grid.length || col >= grid[0].length || grid[row][col] == '0') {
            return;
        }

        grid[row][col] = '0';
        for(int[] dir : directions) {
            dfs(row + dir[0], col + dir[1], grid);
        }
    }
}

File: Data Structures & Algorithms/islands-and-treasure/submission-3.py
class Solution:
    def islandsAndTreasure(self, grid: list[list[int]]) -> None:
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        rows, cols = len(grid), len(grid[0])
        visited = set()
        INF = 2147483647

        def dfs(row: int, col: int) -> int:
            if row < 0 or col < 0 or row >= rows or col >= cols or grid[row][col] == -1 or (row, col) in visited:
                return INF
            if grid[row][col] == 0:
                return 0

            visited.add((row, col))

            result = INF
            for dr, dc in directions:
                distance = dfs(row + dr, col + dc)
                if distance != INF:
                    result = min(result, 1 + distance)
            
            visited.remove((row, col))

            return result
        
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == INF:
                    grid[row][col] = dfs(row, col)